lib/configcat/rolloutevaluator.rb

```diff
@@ -1,197 +1,760 @@
+        if !is_root_flag_evaluation
+          raise e
+        else
+          error = "Failed to evaluate setting '#{key}'. (#{e}). " \
+            "Returning the `%s` parameter that you specified in your application: '#{default_value}'."
+          @log.error(2001, error)
+          return [default_value, default_variation_id, nil, nil, error]
+        end
+      end
+    end
+
+    private
+
+    # Calculates the SHA256 hash of the given value with the given salt and context_salt.
+    def sha256(value_utf8, salt, context_salt)
+      Digest::SHA256.hexdigest(value_utf8 + salt + context_salt)
+    end
+
+    def format_rule(comparison_attribute, comparator, comparison_value)
+      comparator_text = COMPARATOR_TEXTS[comparator]
+      "User.#{comparison_attribute} #{comparator_text} #{EvaluationLogBuilder.trunc_comparison_value_if_needed(comparator, comparison_value)}"
+    end
+
+    def user_attribute_value_to_string(value)
+      return nil if value.nil?
+
+      if value.is_a?(DateTime) || value.is_a?(Time)
+        value = get_user_attribute_value_as_seconds_since_epoch(value)
+      elsif value.is_a?(Array)
+        value = get_user_attribute_value_as_string_list(value)
+        # Convert the array to a JSON string
+        return value.to_json
+      end
+
+      if value.is_a?(Float)
+        return 'NaN' if value.nan?
+        return 'Infinity' if value.infinite? == 1
+        return '-Infinity' if value.infinite? == -1
+        return value.to_s if value.to_s.include?('e')
+        return value.to_i.to_s if value == value.to_i
+      end
+
+      value.to_s
+    end
+
+    def get_user_attribute_value_as_text(attribute_name, attribute_value, condition, key)
+      return attribute_value if attribute_value.is_a?(String)
+
+      @log.warn(3005, "Evaluation of condition (#{condition}) for setting '#{key}' may not produce the expected result " \
+                      "(the User.#{attribute_name} attribute is not a string value, thus it was automatically converted to " \
+                      "the string value '#{attribute_value}'). Please make sure that using a non-string value was intended.")
+      user_attribute_value_to_string(attribute_value)
+    end
+
+    def convert_numeric_to_float(value)
+      if value.is_a?(String)
+        value = value.tr(',', '.')
+        if value == 'NaN'
+          return Float::NAN
+        elsif value == 'Infinity'
+          return Float::INFINITY
+        elsif value == '-Infinity'
+          return -Float::INFINITY
+        end
+      end
+
+      Float(value)
+    end
+
+    def get_user_attribute_value_as_seconds_since_epoch(attribute_value)
+      if attribute_value.is_a?(DateTime) || attribute_value.is_a?(Time)
+        return Utils.get_seconds_since_epoch(attribute_value)
+      end
+
+      convert_numeric_to_float(attribute_value)
+    end
+
+    def get_user_attribute_value_as_string_list(attribute_value)
+      if !attribute_value.is_a?(Array)
```

Comment:
We should try to JSON parse strings only.

```suggestion
      if attribute_value.is_a?(String)
```